src/solids4FoamModels/solidModels/linGeomVelocitySolid/linGeomVelocitySolid.C

```diff
@@ -0,0 +1,244 @@
+    // time-step
+    D() = D().oldTime() + U()*runTime().deltaT();
+
+    // Update gradient of displacement increment
+    mechanical().grad(DD(), gradDD());
+
+    // Update gradient of total displacement
+    gradD() = gradD().oldTime() + gradDD();
+
+    // Calculate the stress using run-time selectable mechanical law
+    mechanical().correct(sigma());
+}
+
+
+// * * * * * * * * * * * * * * * * Constructors  * * * * * * * * * * * * * * //
+
+linGeomVelocitySolid::linGeomVelocitySolid
+(
+    Time& runTime,
+    const word& region
+)
+:
+    solidModel(typeName, runTime, region),
+    impK_(mechanical().impK()),
+    impKf_(mechanical().impKf()),
+    rImpK_(1.0/impK_),
+    predictor_(solidModelDict().lookupOrDefault<Switch>("predictor", false))
+{
+    UisRequired();
+
+    // Force all required old-time fields to be created
+    fvm::ddt(U());
+
+    // For consistent restarts, we will calculate the gradient field
+    U().storePrevIter();
+    D().storePrevIter();
+    mechanical().grad(D(), gradD());
+}
+
+
+// * * * * * * * * * * * * * * * Member Functions  * * * * * * * * * * * * * //
+
+
+bool linGeomVelocitySolid::evolve()
+{
+    Info<< "Evolving solid solver" << endl;
+
+    if (predictor_)
+    {
+        predict();
+    }
+
+    // Mesh update loop
+    do
+    {
+        int iCorr = 0;
+#ifdef OPENFOAMESIORFOUNDATION
+        SolverPerformance<vector> solverPerfU;
+        SolverPerformance<vector>::debug = 0;
+#else
+        lduSolverPerformance solverPerfU;
+        blockLduMatrix::debug = 0;
+#endif
+
+        Info<< "Solving the momentum equation for U" << endl;
+
+        // Momentum equation loop
+        do
+        {
+            // Store fields for under-relaxation and residual calculation
+            U().storePrevIter();
+
+            // Linear momentum equation velocity form
+            fvVectorMatrix UEqn
+            (
+                rho()*fvm::ddt(U())
+             == fvm::laplacian(impKf_*runTime().deltaT(), U(), "laplacian(DD,U)")
+              - fvc::laplacian(impKf_*runTime().deltaT(), U(), "laplacian(DD,U)")
+              + fvc::div(sigma(), "div(sigma)")
+              + rho()*g()
+              + stabilisation().stabilisation(U(), gradU(), impK_*runTime().deltaT())
+            );
+
+            // Under-relaxation the linear system
+            UEqn.relax();
+
+            // Enforce any cell displacements
+            solidModel::setCellDisps(UEqn);
+
+            // Hack to avoid expensive copy of residuals
+#ifdef OPENFOAMESI
+            const_cast<dictionary&>(mesh().solverPerformanceDict()).clear();
+#endif
+
+            // Solve the linear system
+            solverPerfU = UEqn.solve();
+
+            // Fixed or adaptive field under-relaxation
+            relaxField(U(), iCorr);
+            
+            // Update the increment of displacement
+            DD() = U()*runTime().deltaT();
+            // Update the total displacement
```

Comment:
@denMaier, I suggest we add a new line between 151 and 152.

Separately, line 151 uses a 1st order Euler method to update `DD`, i.e. `U = DD/deltaT`. So, if a second-order scheme was used for `fvm::dat(U)`, the final overall scheme will still be first-order in time if the stress depends on `D`. Second-order is possible if the stress directly depends on `U`.

So, what do we do? Some options:
- Leave as it is and optionally give a warning in the constructor if the user selects a second-order scheme for U
- Modify line 151 as follows to allow for first and second order:
```
if (secondOrder)
{
    DD() = U()*runTime().deltaT() + 0.5*a*Foam::pow(runTime().deltaT(), 2);
}
else // first order
{
    DD() = U()*runTime().deltaT();
}
``` 
where `a = fvc::ddt(U)` is the acceleration. There is more than one approach for second order discretisation, so hard-coding this second backward Euler approach could potentially be inconsistent with `fvc::ddt(U)` but maybe that is just a limitation, unless OpenFOAM adds a `fvc::integrate(U)` option.

Reply:
Ah yes.. Thanks @philipcardiff,  you are right. I had thought about that before, but it fell through while coding.
Do you know if there are any plans for a fvc::integrate(U) option?

Might be a silly idea, but wouldn't it be possible to somehow use fvm::ddt(D)=U and solve with a diagonal solver?

Reply:
I don't know of any plans for `fvc::integrate(U)`; I imagine it is unlikely, as it would require each ddt discretisation to give its inverse.

That's an interesting idea to use `fvm::ddt(D)=U`... I guess that could work; I think OpenFOAM will automatically use the diagonal solver since the system is diagonal. Of course, in this case, it is the user's responsibility they select ddt schemes for D and U that are compatible, but that's fine. Feel free to check that this works and push the change.

Reply:
fvm::ddt(D)==U works nicely for the internalField and hence calculating the stress field. I'm not quite sure, but it seems like the boundary values are not calculated this way, which is a pitty. Hopefully, i can find some way to update them accordingly. 

Further, I noticed, that I still use first order Euler twice in my code and I'm not quite sure how to overcome this.
The first is in the laplacian terms. I'm not 100% sure how bad it is, since they should vanish with convergence, however I'm pretty sure convergence rates might be sub-optimal this way.
```
fvVectorMatrix UEqn
            (
                rho()*fvm::ddt(U())
             == fvm::laplacian(impKf_*runTime().deltaT(), U(), "laplacian(DD,U)")
              - fvc::laplacian(impKf_*runTime().deltaT(), U(), "laplacian(DD,U)")
              + fvc::div(sigma(), "div(sigma)")
              + rho()*g()
              + stabilisation().stabilisation(U(), gradU(), impK_*runTime().deltaT())
            );
```
Second time is in boundary gradient calculation for the force boundary condition:
```
                (traction - n*pressure)
              - (n & (pSigma - pImpK*runTime().deltaTValue()*pGradU))
            )*pRImpK/runTime().deltaTValue()
```

Reply:
> fvm::ddt(D)==U works nicely for the internalField and hence calculating the stress field. I'm not quite sure, but it seems like the boundary values are not calculated this way, which is a pitty. Hopefully, i can find some way to update them accordingly.

Hmnn... after this equation is solved, `D.correctionBoundaryConditions()` will be automatically called. Normally, `D` will have boundary conditions like `fixedDisplacement` and `solidTraction`, which correctly set the boundary values; however, here, presumably `D` has `calculated` boundary conditions, which do nothing and will stay as zero.

So what do we do... we can't use `zeroGradient` as this will result in unphysical values for gradD and strain near the boundary. We could expect that the user gives `D` in `0` along with `U`, where the `D` boundary conditions are compatible with the `U` conditions (this is somewhat inelegant and open for mistakes). We could write a procedure to extrapolate the boundary displacements from the internal field. Or we could go back to the `if (firstOrder) ...` solution.
 
> 
> Further, I noticed, that I still use first order Euler twice in my code and I'm not quite sure how to overcome this. The first is in the laplacian terms. I'm not 100% sure how bad it is, since they should vanish with convergence, however I'm pretty sure convergence rates might be sub-optimal this way.
> 
> ```
> fvVectorMatrix UEqn
>             (
>                 rho()*fvm::ddt(U())
>              == fvm::laplacian(impKf_*runTime().deltaT(), U(), "laplacian(DD,U)")
>               - fvc::laplacian(impKf_*runTime().deltaT(), U(), "laplacian(DD,U)")
>               + fvc::div(sigma(), "div(sigma)")
>               + rho()*g()
>               + stabilisation().stabilisation(U(), gradU(), impK_*runTime().deltaT())
>             );
> ``` 

I think this is fine; this will only affect convergence and not the final answer.

> Second time is in boundary gradient calculation for the force boundary condition:
> 
> ```
>                 (traction - n*pressure)
>               - (n & (pSigma - pImpK*runTime().deltaTValue()*pGradU))
>             )*pRImpK/runTime().deltaTValue()
> ```

I think this is similar to the Laplacian terms in the moment equation, as this equation is derived by adding and taking away the same (or a similar) term. So at convergence, the effect of `pImpK*runTime().deltaTValue()*pGradU)` disappears and instead $\boldsymbol{t} = \boldsymbol{n} \cdot \boldsymbol{\sigma}$ is enforced. So, this is fine too.